File: 13-Algorithms_sorting/sorting/lib/merge_sort.ex
defmodule MergeSort do
  def sort([]), do: []
  def sort([x]), do: [x]
  def sort(lst) when not is_list(lst), do: nil

  def sort(lst) do
    mid = div(Enum.count(lst), 2)
    {left, right} = Enum.split(lst, mid)
    merge(sort(left), sort(right))
  end

  defp merge(left, []), do: left
  defp merge([], right), do: right

  defp merge([lh | lt], right = [rh | _]) when lh <= rh do
    [lh | merge(lt, right)]
  end

  defp merge(left, [rh | rt]), do: [rh | merge(left, rt)]
end

# MergeSort.sort([99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0])

File: 13-Algorithms_sorting/merge_sort.rb
# frozen_string_literal: true

class MergeSort
  def sort(arr)
    return arr if arr.length < 2

    # Split array in into right and left
    mid = arr.length / 2
    left = arr.take(mid)
    right = arr.drop(mid)

    merge(sort(left), sort(right))
  end

  private

  def merge(left, right)
    return right if left.empty?
    return left if right.empty?

    result = if left.first <= right.first
               left.shift
             else
               right.shift
             end

    [result] + merge(left, right)
  end

  # def merge(left, right)
  #   result = []

  #   while !left.empty? && !right.empty?

  #     result << if left.first <= right.first
  #                 left.shift
  #               else
  #                 right.shift
  #               end
  #   end

  #   result.concat(left, right)
  # end
end

numbers = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0]
m = MergeSort.new
p m.sort(numbers)
